### ISSUE TITLE
systemd 228: IPv6AcceptRouterAdvertisements=yes or unset accepts too many prefixes

### ISSUE THREAD

**Issue by mikken** [systemd/systemd#2004]
This is the issue I got after updating systemd from 226 to 228.
I have a few interfaces managed by systemd-networkd. Two of them should accept RA (SLAAC mode).
If I leave IPv6AcceptRouterAdvertisements unset or set it to yes, _every_ advertised prefix is applied to _every_ interface. This results in every interface having its normal global IPv6 addr plus one weird address from another network.
If I set IPv6AcceptRouterAdvertisements=no on all of them and use the good old accept_ra=2 setting (written by a oneshot service after systemd-networkd starts), everything works as expected (every interface gets one global IPv6 addr each).

**Comment by mikken** [comment]
By the way, I got an impression that "yes" and "no" settings are too few settings for this option. It's not flexible enough.
If I leave it unset, RAs are still managed by systemd-networkd if IPForward is unset.
What should I do if I want to leave SLAAC address management to the kernel? Set IPv6AcceptRouterAdvertisements=no and manage accept_ra= in a different way. This is unnatural.
Something like IPv6AcceptRouterAdvertisements=kernel would be a good choice.

**Comment by mikken** [comment]
One more related comment: I set accept_ra_defrtr=0 to ignore default route for an interface and only accept the address and local route. There's no way I can do it with systemd-networkd currently.

**Comment by brain0** [comment]
I have the same experience. I could not get my network setup to work with systemd 228 at all - I tried setting all ipv6 related options to =kernel as a workaround, but that resulted in ipv6 being entirely disabled.

All other ipv6 options also behave completely unexpected - for example, privacy extensions are never enabled.

**Comment by eworm-de** [comment]
I do see this as well:

```
# ip addr
[...]
4: wl: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
[...]
    inet6 2a01:4f8:150:xxxx:xxxx:xxxx:xxxx:xxxx/64 scope global temporary dynamic 
       valid_lft 602873sec preferred_lft 83873sec
[...]
5: en: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
[...]
    inet6 2a01:4f8:150:xxxx:xxxx:xxxx:xxxx:xxxx/64 scope global noprefixroute dynamic 
       valid_lft 2591891sec preferred_lft 604691sec
[...]
```

Things work as long as the system is up.
It's even worse after suspend and resume: I have a stale address on the _wrong_ interface, the correct interface does not receive an IPv6 address. IPv6 traffic is routed to a black hole, resulting in bad timeouts until connections fall back to IP.

**Comment by eworm-de** [comment]
Looks like this was introduced with commit 3b015d40 by @teg. `ndisc_prefix_autonomous_handler()` is called for interfaces that did not receive a RA.

**Comment by RoadRunnr** [comment]
@eworm-de close, but's it seems that real problem is the binding of the multicast socket to the device. Under Linux 4.4 is get this with strace:

first multicast socket is bound to first interface:

```
441   socket(PF_INET6, SOCK_RAW|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_ICMPV6) = 9
441   setsockopt(9, SOL_ICMPV6, 1, "\377\377\377\377\377\377\377\377\377\377\377\377\377\377\377\377\277\377\377\377\377\377\377\377\377\377\377\377\377\377\377\377", 32) = 0
441   setsockopt(9, SOL_IPV6, IPV6_MULTICAST_IF, [7], 4) = 0
441   setsockopt(9, SOL_IPV6, IPV6_MULTICAST_LOOP, [0], 4) = 0
441   setsockopt(9, SOL_IPV6, IPV6_MULTICAST_HOPS, [255], 4) = 0
441   setsockopt(9, SOL_IPV6, IPV6_ADD_MEMBERSHIP, "\377\2\0\0\0\0\0\0\0\0\0\0\0\0\0\1\7\0\0\0", 20) = 0
441   setsockopt(9, SOL_IPV6, 0x33 /* IPV6_??? */, [1], 4) = 0
```

second multicast socket is bound to second interface:

```
441   socket(PF_INET6, SOCK_RAW|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_ICMPV6) = 11
441   setsockopt(11, SOL_ICMPV6, 1, "\377\377\377\377\377\377\377\377\377\377\377\377\377\377\377\377\277\377\377\377\377\377\377\377\377\377\377\377\377\377\377\377", 32) = 0
441   setsockopt(11, SOL_IPV6, IPV6_MULTICAST_IF, [2], 4) = 0
441   setsockopt(11, SOL_IPV6, IPV6_MULTICAST_LOOP, [0], 4) = 0
441   setsockopt(11, SOL_IPV6, IPV6_MULTICAST_HOPS, [255], 4) = 0
441   setsockopt(11, SOL_IPV6, IPV6_ADD_MEMBERSHIP, "\377\2\0\0\0\0\0\0\0\0\0\0\0\0\0\1\2\0\0\0", 20) = 0
441   setsockopt(11, SOL_IPV6, 0x33 /* IPV6_??? */, [1], 4) = 0
```

So far, everything is fine. But now a answer arrives:

```
441   recvmsg(11, {msg_name(28)={sa_family=AF_INET6, sin6_port=htons(0), inet_pton(AF_INET6, "fe80::250:56ff:fe96:a186", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=if_nametoindex("ge0p5")}, msg_iov(1)=[{"\206\0\246\35@\0\7\10\0\0\0\0\0\0\0\0\1\1\0PV\226\241\206\5\1\0\0\0\0\5\334", 32}], msg_controllen=24, {cmsg_len=20, cmsg_level=SOL_IPV6, cmsg_type=, ...}, msg_flags=0}, 0) = 32
...
441   recvmsg(9, {msg_name(28)={sa_family=AF_INET6, sin6_port=htons(0), inet_pton(AF_INET6, "fe80::250:56ff:fe96:a186", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=if_nametoindex("ge0p5")}, msg_iov(1)=[{"\206\0\246\35@\0\7\10\0\0\0\0\0\0\0\0\1\1\0PV\226\241\206\5\1\0\0\0\0\5\334", 32}], msg_controllen=24, {cmsg_len=20, cmsg_level=SOL_IPV6, cmsg_type=, ...}, msg_flags=0}, 0) = 32
```

WTF??? Sockets should be properly bound to the interfaces, but recvmsg does get the same message on **BOTH** sockets.

Clearly the socket to interface binding is not working as expected.

A quick workaround might be to filter sin6_scope_id against the expected ifindex.

**Comment by RoadRunnr** [comment]
After some experiments I've come up with a temporary fix:

```
diff --git a/src/libsystemd-network/sd-ndisc.c b/src/libsystemd-network/sd-ndisc.c
index d8154f0..643391e 100644
--- a/src/libsystemd-network/sd-ndisc.c
+++ b/src/libsystemd-network/sd-ndisc.c
@@ -531,6 +531,9 @@ static int ndisc_router_advertisment_recv(sd_event_source *s, int fd, uint32_t r
         else if (msg.msg_namelen != sizeof(sa.in6)) {
                 log_ndisc(nd, "Received invalid source address size from ICMPv6 socket: %zu bytes", (size_t)msg.msg_namelen);
                 return 0;
+        } else if (sa.in6.sin6_scope_id != nd->index) {
+                log_ndisc(nd, "Received RA on wrong interface: %d != %d", sa.in6.sin6_scope_id, nd->index);
+                return 0;
         } else
                 gw = &sa.in6.sin6_addr;

```

The underlying issue is that the code assume that we can bind a multicast socket to receive on a specific interface only. It uses the IPV6_MULTICAST_IF socket option for that. Problem is, that option only sets the out interface and does not filter on receive.

I tried to bind with sin6_scope_id set to the interface, it does return success, but doesn't have the desired effect either.

SO_BINDTODEVICE might be another option, but I didn't try that.

@teg any idea for better fix?

**Comment by eworm-de** [comment]
@RoadRunnr Has your change been committed to any repository? I would like to cherry-pick and give it a try.

**Comment by eworm-de** [comment]
BTW, from `man 7 socket` about socket option `SO_BINDTODEVICE`:

```
Note that this works only for some socket types, particularly AF_INET
sockets.  It is not supported for packet sockets (use normal bind(2) there).
```

**Comment by RoadRunnr** [comment]
@eworm-de No, I didn't push this any repository. But it's only three lines, applying that manually to the source shouldn't be too difficult.

I also see this as band-aid only. The real fix would IMHO be either to figure out how to bind multicast sockets to interfaces for receiving or rewrite the multicast logic to use only one socket.

**Comment by RoadRunnr** [comment]
@eworm-de I tried the bind approach, but that didn't work. The bind call returned ok, but I was still receiving the message on the wrong socket.

**Comment by mikken** [comment]
Has anyone tested with v229?

**Comment by brain0** [comment]
229 is still as broken as 228.

**Comment by smpolymen** [comment]
I think something similar to @RoadRunnr 's proposed band-aid is the only real way to get this to work. 

Another option would be to set IPV6_PKTINFO in icmp6_bind_router_solicitation and use the interface index directly from the in6_pktinfo struct, but you still have the issue of getting the same message N times. 

There doesn't seem to be a way around IPV6_ADD_MEMBERSHIP causing all interfaces in the group to receive, as that seems to be the intended functionality.

**Comment by mmonaco** [comment]
@RoadRunnr Thanks! Your patch helped me.

**Comment by ssahani** [comment]
fix is #3294 .
